core/src/main/java/kafka/server/NetworkPartitionMetadataClient.java

```diff
@@ -0,0 +1,287 @@
+                tp.topic(),
+                tp.partition(),
+                listenerName
+            );
+
+            if (leaderNodeOpt.isEmpty() || leaderNodeOpt.get().isEmpty()) {
+                // No leader available - complete with error
+
+                future.complete(new OffsetResponse(-1, Errors.LEADER_NOT_AVAILABLE));
+                continue;
+            }
+
+            Node leaderNode = leaderNodeOpt.get();
+            partitionsByNode.computeIfAbsent(leaderNode, k -> new ArrayList<>()).add(tp);
+        }
+
+        // Create and enqueue requests for each node
+        for (Map.Entry<Node, List<TopicPartition>> entry : partitionsByNode.entrySet()) {
+            Node node = entry.getKey();
+            List<TopicPartition> partitions = entry.getValue();
+
+            // Create a map of futures only for partitions in this request
+            Map<TopicPartition, CompletableFuture<OffsetResponse>> partitionFutures = new HashMap<>();
+            for (TopicPartition tp : partitions) {
+                partitionFutures.put(tp, futures.get(tp));
+            }
+
+            // Create ListOffsetsRequest for this node
+            ListOffsetsRequest.Builder requestBuilder = createListOffsetsRequest(partitions);
+
+            // Create pending request to track this request
+            PendingRequest pendingRequest = new PendingRequest(node, partitions, partitionFutures, requestBuilder);
+
+            // Enqueue to send thread
+            sendThread.enqueue(pendingRequest);
+        }
+
+        return futures;
+    }
+
+    /**
+     * Creates a ListOffsetsRequest Builder for the given partitions requesting latest offsets.
+     */
+    private ListOffsetsRequest.Builder createListOffsetsRequest(List<TopicPartition> partitions) {
+        // Group partitions by topic name
+        Map<String, List<TopicPartition>> partitionsByTopic = partitions.stream()
+            .collect(Collectors.groupingBy(TopicPartition::topic));
+
+        List<ListOffsetsTopic> topics = new ArrayList<>();
+        for (Map.Entry<String, List<TopicPartition>> entry : partitionsByTopic.entrySet()) {
+            String topicName = entry.getKey();
+            ListOffsetsTopic topic = new ListOffsetsTopic().setName(topicName);
+
+            for (TopicPartition tp : entry.getValue()) {
+                topic.partitions().add(
+                    new ListOffsetsPartition()
+                        .setPartitionIndex(tp.partition())
+                        .setTimestamp(ListOffsetsRequest.LATEST_TIMESTAMP)
+                        .setCurrentLeaderEpoch(-1) // Will be set by broker if needed
+                );
+            }
+            topics.add(topic);
+        }
+
+        // Isolation level will always be READ_UNCOMMITTED when finding the partition end offset.
+        return ListOffsetsRequest.Builder.forConsumer(
+            true,
+            IsolationLevel.READ_UNCOMMITTED
+        ).setTargetTimes(topics);
```

Comment:
Similar but avoids `new Arraylist` in Collectors.groupingBy

 
```suggestion
        Map<String, ListOffsetsTopic> topicsMap = new HashMap<>();
        partitions.forEach(tp -> {
            if (!topicsMap.containsKey(tp.topic())) {
                ListOffsetsTopic topic = new ListOffsetsTopic().setName(tp.topic());
                topicsMap.put(tp.topic(), topic);
            }
            ListOffsetsTopic topic = topicsMap.get(tp.topic());
            topic.partitions().add(
                new ListOffsetsPartition()
                    .setPartitionIndex(tp.partition())
                    .setTimestamp(ListOffsetsRequest.LATEST_TIMESTAMP)
                    .setCurrentLeaderEpoch(-1) // Will be set by broker if needed
            );
        });

        // Isolation level will always be READ_UNCOMMITTED when finding the partition end offset.
        return ListOffsetsRequest.Builder.forConsumer(
            true,
            IsolationLevel.READ_UNCOMMITTED
        ).setTargetTimes(List.copyOf(topicsMap.values()));
```